JAICore/jaicore-ml/src/jaicore/ml/cache/ReproducibleInstances.java

```diff
@@ -0,0 +1,175 @@
+package jaicore.ml.cache;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import jaicore.ml.WekaUtil;
+import jaicore.ml.openml.OpenMLHelper;
+import weka.core.Instances;
+
+/**
+ * New Instances class to track splits and data origin.
+ * 
+ * @author jnowack
+ * @author mirko
+ * @author jonas
+ *
+ */
+public class ReproducibleInstances extends Instances {
+
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = 4318807226111536282L;
+	private List<Instruction> history = new LinkedList<>();
+	private boolean cacheStorage = true;
+	private boolean cacheLookup = true;
+	
+	private ReproducibleInstances(Instances dataset) {
+		super(dataset);
+	}
+	
+	/**
+	 * Creates a {@link ReproducibleInstances} Object based on the given History. Instructions that no not modify the Instances will be ignored (No evaluation will be done). 
+	 * 
+	 * @param history - List of INstructions used to create the original Instances
+	 * @param apiKey - apiKey in case openml.org is used
+	 * @return new {@link ReproducibleInstances} object
+	 * @throws IOException 
+	 * @throws  
+	 */
+	public static ReproducibleInstances fromHistory(List<Instruction> history, String apiKey) throws IOException {
+		ReproducibleInstances instances = null;
+		for (int i = 0; i < history.size(); i++) {
+			Instruction inst = history.get(i);
+			switch (inst.command) {
+			case "loadDataset":
+				// load openml or local dataset
+				if(inst.inputs.get("provider").equals("openml.org")) {
+					instances = fromOpenML(inst.inputs.get("id"), apiKey);
+				}
+				else if(inst.inputs.get("provider").startsWith("local")) {
+					instances = fromARFF(inst.inputs.get("id"), inst.inputs.get("provider").split("\\.")[1]);
+				}
+				break;
+			case "split":
+				// create split
+				String[] ratiosAsStrings = inst.getInputs().get("ratios").split(",");
+				double[] ratios = new double[ratiosAsStrings.length];
+				for (int j = 0; j < ratiosAsStrings.length; j++) {
+					ratios[j] = Double.parseDouble(ratiosAsStrings[j].trim().substring(1, ratiosAsStrings[j].length()-1));
+				}
+				instances = WekaUtil.getStratifiedSplit(instances, Long.parseLong(inst.inputs.get("seed")), ratios).get( Integer.parseInt(inst.getInputs().get("outIndex")));
+				break;
+			default:
+				break;
+			}
+		}
+		return instances;
+	}
+	
+
+	
+	public ReproducibleInstances(ReproducibleInstances dataset) {
+		super(dataset);
+		for (Iterator<Instruction> iterator = dataset.history.iterator(); iterator.hasNext();) {
+			Instruction i = iterator.next();
+			history.add(i);
+		}
+		cacheLookup = dataset.cacheLookup;
+		cacheStorage = dataset.cacheStorage;
+	}
+
+	/**
+	 * Creates a new {@link ReproducibleInstances} object. Data is loaded from
+	 * openml.org.
+	 * 
+	 * @param id     The id of the openml dataset
```

Comment:
either start with a small letter or with a large one but do it consistently among all files